be/src/exec/rowid_fetcher.cpp

```diff
@@ -275,4 +284,197 @@ Status RowIDFetcher::fetch(const vectorized::ColumnPtr& column_row_ids,
+    // read from storage engine row id by row id
+    OlapReaderStatistics stats;
+    vectorized::Block result_block;
+    int64_t acquire_tablet_ms = 0;
+    int64_t acquire_rowsets_ms = 0;
+    int64_t acquire_segments_ms = 0;
+    int64_t lookup_row_data_ms = 0;
+
+    // init desc
+    TupleDescriptor desc(request.desc());
+    std::vector<SlotDescriptor> slots;
+    slots.reserve(request.slots().size());
+    for (const auto& pslot : request.slots()) {
+        slots.push_back(SlotDescriptor(pslot));
+        desc.add_slot(&slots.back());
+    }
+
+    // init read schema
+    TabletSchema full_read_schema;
+    for (const ColumnPB& column_pb : request.column_desc()) {
+        full_read_schema.append_column(TabletColumn(column_pb));
+    }
+
+    std::unordered_map<IteratorKey, IteratorItem, HashOfIteratorKey> iterator_map;
+    // read row by row
+    for (size_t i = 0; i < request.row_locs_size(); ++i) {
+        const auto& row_loc = request.row_locs(i);
+        MonotonicStopWatch watch;
+        watch.start();
+        BaseTabletSPtr tablet = scope_timer_run(
+                [&]() {
+                    if (!config::is_cloud_mode()) {
+                        auto tablet =
+                                static_cast<StorageEngine*>(_engine)->tablet_manager()->get_tablet(
+                                        row_loc.tablet_id(), true /*include deleted*/);
+                        return std::dynamic_pointer_cast<BaseTablet>(tablet);
+                    } else {
+                        auto res =
+                                static_cast<CloudStorageEngine*>(_engine)->tablet_mgr().get_tablet(
+                                        row_loc.tablet_id(), true /*include deleted*/);
+                        return std::dynamic_pointer_cast<BaseTablet>(res.value());
+                    }
+                },
+                &acquire_tablet_ms);
+        RowsetId rowset_id;
+        rowset_id.init(row_loc.rowset_id());
+        if (!tablet) {
+            continue;
+        }
+        // We ensured it's rowset is not released when init Tablet reader param, rowset->update_delayed_expired_timestamp();
+        BetaRowsetSharedPtr rowset = std::static_pointer_cast<BetaRowset>(scope_timer_run(
+                [&]() { return _engine->get_quering_rowset(rowset_id); }, &acquire_rowsets_ms));
+        if (!rowset) {
+            LOG(INFO) << "no such rowset " << rowset_id;
+            continue;
+        }
+        size_t row_size = 0;
+        Defer _defer([&]() {
+            LOG_EVERY_N(INFO, 100)
+                    << "multiget_data single_row, cost(us):" << watch.elapsed_time() / 1000
+                    << ", row_size:" << row_size;
+            *response->add_row_locs() = row_loc;
+        });
+        SegmentCacheHandle segment_cache;
+        RETURN_IF_ERROR(scope_timer_run(
+                [&]() {
+                    return SegmentLoader::instance()->load_segments(rowset, &segment_cache, true);
```

Comment:
Is SegmentLoader the same for both cloud and local mode?

Reply:
they are same